src/hotspot/share/runtime/threads.cpp

```diff
@@ -1194,7 +1194,11 @@ GrowableArray<JavaThread*>* Threads::get_pending_threads(ThreadsList * t_list,
     // The first stage of async deflation does not affect any field
     // used by this comparison so the ObjectMonitor* is usable here.
     address pending = (address)p->current_pending_monitor();
-    if (pending == monitor) {             // found a match
+    address waiting = (address)p->current_waiting_monitor();
+    if (pending == monitor ||
+        (waiting == monitor && JavaThreadStatus::BLOCKED_ON_MONITOR_ENTER ==
+         java_lang_Thread::get_thread_status(p->vthread_or_thread()))
```

Comment:
This code seems extremely racy. Is there anything that is stopping the target thread from running while we attempt this inspection of the oop state? If it was waiting and was interrupted or timed-out then we could mistakenly think it was still pending entry to this monitor when in fact it is not.
It is also a shame that we have to reach up into the oop thread state to figure out if if it is blocked on re-entry ...

Reply:
The function `get_pending_threads()` is executed only by the VM_GetObjectMonitorUsage operation.
I've added this assert there: 
    `assert(Thread::current()->is_VM_thread(), "Must be the VM thread");`
Does it address your concern?

Reply:
Yes thanks. Didn't realize we were in a safepoint VM op.